src/io/flutter/run/daemon/DevToolsService.java

```diff
@@ -58,7 +58,19 @@ public static DevToolsService getInstance(@NotNull final Project project) {
 
   private DevToolsService(@NotNull final Project project) {
     this.project = project;
+  }
+
+  public CompletableFuture<DevToolsInstance> getDevToolsInstance() {
+    // Create instance if it doesn't exist yet, or if the previous attempt failed.
+    if (devToolsInstance == null || devToolsInstance.isCompletedExceptionally()) {
+      devToolsInstance = new CompletableFuture<>();
+      startServer();
```

Comment:
Consider passing the devToolsInstance as a parameter to startServer() so that we always complete the right future (given that this field is reused over time for new instances).

Reply:
I'm not sure if this fully addresses the problem, but I put the future into an atomic reference to ensure that there can only be one at a time (one future, and therefore one process / process handler). It could be problematic if more than one server gets started and the reference to shut it down on close gets lost.